libminifi/src/utils/SystemCPUUsageTracker.cpp

```diff
@@ -0,0 +1,190 @@
+
+  return percent;
+}
+#endif  // linux
+
+#ifdef WIN32
+SystemCPUUsageTracker::SystemCPUUsageTracker() :
+    total_idle_(0), total_sys_(0), total_user_(0),
+    previous_total_idle_(0), previous_total_sys_(0), previous_total_user_(0) {
+  queryCPUTimes();
+}
+
+double SystemCPUUsageTracker::getCPUUsageAndRestartCollection() {
+  queryCPUTimes();
+  if (isCurrentQuerySameAsPrevious() || isCurrentQueryOlderThanPrevious()) {
+    return -1.0;
+  } else {
+    return getCPUUsageBetweenLastTwoQueries();
+  }
+}
+
+void SystemCPUUsageTracker::queryCPUTimes() {
+  previous_total_user_ = total_user_;
+  previous_total_sys_ = total_sys_;
+  previous_total_idle_ = total_idle_;
+  FILETIME fidle, fsys, fuser;
+  GetSystemTimes(&fidle, &fsys, &fuser);
+  total_user_ = ULARGE_INTEGER{ fuser.dwLowDateTime, fuser.dwHighDateTime }.QuadPart;
+  total_sys_ = ULARGE_INTEGER{ fsys.dwLowDateTime, fsys.dwHighDateTime }.QuadPart;
+  total_idle_ = ULARGE_INTEGER{ fidle.dwLowDateTime, fidle.dwHighDateTime }.QuadPart;
+}
+
+bool SystemCPUUsageTracker::isCurrentQueryOlderThanPrevious() const {
+  return (total_user_ < previous_total_user_ ||
+    total_sys_ < previous_total_sys_ ||
+    total_idle_ < previous_total_idle_);
+}
+
+bool SystemCPUUsageTracker::isCurrentQuerySameAsPrevious() const {
+  return (total_user_ == previous_total_user_ &&
+    total_sys_ == previous_total_sys_ &&
+    total_idle_ == previous_total_idle_);
+}
+
+double SystemCPUUsageTracker::getCPUUsageBetweenLastTwoQueries() const {
+  uint64_t total_user_diff = total_user_ - previous_total_user_;
+  uint64_t total_sys_diff = total_sys_ - previous_total_sys_;
+  uint64_t total_idle_diff = total_idle_ - previous_total_idle_;
+  uint64_t total_diff = total_user_diff + total_sys_diff;
+  double percent = static_cast<double>(total_diff - total_idle_diff) / static_cast<double>(total_diff);
+
+  return percent;
+}
+#endif  // windows
+
+#ifdef __APPLE__
+SystemCPUUsageTracker::SystemCPUUsageTracker() :
+    total_ticks_(0), previous_total_ticks_(0),
+    idle_ticks_(0), previous_idle_ticks_(0) {
+  queryCPUTicks();
+}
+
+double SystemCPUUsageTracker::getCPUUsageAndRestartCollection() {
+  queryCPUTicks();
+  if (isCurrentQuerySameAsPrevious() || isCurrentQueryOlderThanPrevious()) {
+    return -1.0;
+  } else {
+    return getCPUUsageBetweenLastTwoQueries();
+  }
+}
+
+void SystemCPUUsageTracker::queryCPUTicks() {
+  host_cpu_load_info_data_t cpuinfo;
+  mach_msg_type_number_t count = HOST_CPU_LOAD_INFO_COUNT;
+  auto query_result = host_statistics(mach_host_self(), HOST_CPU_LOAD_INFO, (host_info_t)&cpuinfo, &count);
+  if (query_result == KERN_SUCCESS) {
+    previous_total_ticks_ = total_ticks_;
+    previous_idle_ticks_ = idle_ticks_;
+    total_ticks_ = 0;
+    for (int i = 0; i < CPU_STATE_MAX; i++) {
+      total_ticks_ += cpuinfo.cpu_ticks[i];
+    }
+    idle_ticks_ = cpuinfo.cpu_ticks[CPU_STATE_IDLE];
+  }
```

Comment:
I can't verify this logic because it's incredibly hard to find proper documentation for [`host_statistics`](https://developer.apple.com/documentation/kernel/1502546-host_statistics). Could you explain how you came up with the solution?

Reply:
Yeah, apples cpp documentation is not the best. The basic idea is the same as in the other platforms, I've got the idea to use host_statistics from forums, but managed to verify it: Apple uses this in very similar fashion:
https://github.com/apple/foundationdb/blob/6571e9e78cd0f22af5e283f0d5b7c3dde661afc4/flow/Platform.actor.cpp#L1177